pkg/image/registry/imagestream/etcd/image.go

```diff
@@ -0,0 +1,222 @@
+package etcd
+
+import (
+	"fmt"
+
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/client-go/tools/cache"
+
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+)
+
+// ImageLayerIndex is a cache of image digests to the layers they contain.
+// Because a very large number of images can exist on a cluster, we only
+// hold in memory a small subset of the full image object.
+type ImageLayerIndex interface {
+	HasSynced() bool
+	GetByKey(key string) (item interface{}, exists bool, err error)
+	Run(stopCh <-chan struct{})
+}
+
+type ImageStore interface {
+	rest.Watcher
+	rest.Lister
+}
+
+type imageLayerIndex struct {
+	informer cache.SharedIndexInformer
+}
+
+func NewEmptyLayerIndex() ImageLayerIndex {
+	return imageLayerIndex{}
+}
+
+func (i imageLayerIndex) HasSynced() bool {
+	if i.informer == nil {
+		return true
+	}
+	return i.informer.HasSynced()
+}
+func (i imageLayerIndex) GetByKey(key string) (item interface{}, exists bool, err error) {
+	if i.informer == nil {
+		return nil, false, nil
+	}
+	return i.informer.GetStore().GetByKey(key)
+}
+func (i imageLayerIndex) Run(stopCh <-chan struct{}) {
+	if i.informer == nil {
+		return
+	}
+	i.informer.Run(stopCh)
+}
+
+// NewImageLayerIndex creates a new index over a store that must return
+// images.
+func NewImageLayerIndex(store ImageStore) ImageLayerIndex {
+	ctx := apirequest.NewContext()
+	informer := cache.NewSharedIndexInformer(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			obj, err := store.List(ctx, &metainternalversion.ListOptions{
```

Comment:
yeah, you can do this via a client.  You're creating your own lister/watcher anyway.